### PR TITLE
GC Tombstone: Add allowTombstone request header to enable apps to recover documents from GC-blocking bugs

### PR DESCRIPTION
**Pull request by markfields**
## Description

tl;dr - Adds support for a request header `allowTombstone` that, when passed as true, will enable callers to request a tombstoned object without failing.

_Context: GC Tombstones are objects that GC knows are safe to delete, but rather than deleting them we just mark them as such. This "Tombstone Phase" of rolling out GC allows an app to confirm that there are no GC-blocking bugs in production without exposing their users to permanent data loss._

In order for a host to effectively recover when a GC-impacting bug is found during the Tombstone phase of GC, we're adding a way (via a request header) for app code to request a tombstoned object and receive it back without error.  At this point they can repair whatever incorrect reference was in play that cause the object to be wrongly tombstoned, and in subsequent summaries the object will no longer be a tombstone and the document will be fine.

Fixes [AB#2933](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/2933)

## Reviewer Guidance

...

### CODE REVIEW

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 1]
```diff
@@ -268,16 +268,23 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
-                url: request.url,
-            };
-            sendGCTombstoneEvent(this.mc, event, this.runtime.clientDetails.type === summarizerClientType, [BlobManager.basePath], error);
-            if (this.throwOnTombstoneUsage) {
+            sendGCTombstoneEvent(
+                this.mc.logger,
+                {
+                    eventName: "GC_Tombstone_Blob_Requested",
+                },
+                shouldFail /* logAsError */,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/blobManager.ts:282]: We should also log whether `allowTombstone` was set to give us an idea of when this called for recovery purposes.

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 2]
```diff
@@ -268,16 +268,23 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
+                this.runtime.clientDetails.type === summarizerClientType,
+                [BlobManager.basePath],
+                error,
+            );
+            if (shouldFail) {
                 throw error;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/blobManager.ts:288]: This error will be returned by handle.get(). We need to make sure that consumers can look at it and know that this is caused by tombstone and re-try with the header.

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 3]
```diff
@@ -430,33 +430,35 @@ export class DataStores implements IDisposable {
...
+
+            // Note: if a user writes a request to look like it's viaHandle,
+            // it will be indistinguishable in telemetry from an actual handle
             sendGCTombstoneEvent(
-                this.mc,
-                event,
+                this.mc.logger,
+                {
+                    eventName: "GC_Tombstone_DataStore_Requested",
+                    viaHandle: headerData.viaHandle,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/dataStores.ts:453]: Let's just log `...headerData` so we can also tell if allowTombstone as set or not.

File: packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts [hunk 4]
```diff
@@ -3,29 +3,21 @@
...
+    logger: ITelemetryLogger,
     event: ITelemetryGenericEvent,
+    logAsError: boolean,
     isSummarizerClient: boolean,
     packagePath: readonly string[] | undefined,
-    error?: any,
+    error: unknown,
 ) {
-    const throwOnTombstoneUsage = event.throwOnTombstoneUsage = mc.config.getBoolean(throwOnTombstoneUsageKey) ?? false;
+    event.category = logAsError ? "error" : "generic";
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollectionTombstoneUtils.ts:18]: nit: It bugs me that this function accepts event along with properties that are added to event. It would be better to pass in all the properties directly and create an event in this function.

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 5]
```diff
@@ -775,20 +775,19 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
...
+                    eventName: "GC_Tombstone_DataStore_Changed",
+                    callSite,
+                },
+                this.throwOnTombstoneUsage /* logAsError */,
+                this.clientDetails.type === summarizerClientType,
+                this.pkg,
+                error,
+            );
             if (this.throwOnTombstoneUsage) {
                 throw error;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:792]: We should update this to not throw right? Otherwise, even if data stores are revived by using the recovery scenario, using them will result in this error being thrown. Do you plan to do that in a separate PR?

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 6]
```diff
@@ -507,13 +531,14 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
-            const request: IRequest = { url: unreferencedId, headers: { [RuntimeHeaders.viaHandle]: true } };
-            const response = await dataObject._context.IFluidHandleContext.resolveHandle(request);
+            const serializer = new FluidSerializer(dataObject._context.IFluidHandleContext, () => {});
+            const handle: IFluidHandle = parseHandles({ type: "__fluid_handle__", url: unreferencedId }, serializer);
 
-            assert(response !== undefined, `Expecting a response!`);
-            assert(response.status === 404);
-            assert(response.value === `Datastore removed by gc: ${unreferencedId}`);
+            // Note: if a user makes a request that looks like this, we will also think that the request is via handle
+            await assert.rejects(
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:538]: What about recovery in this case? It could be our bug or app's but in either case, they should be able to recover right?

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 5 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:792]: Yeah gonna raise/merge that PR first

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 2 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/blobManager.ts:288]: I'm gonna revert blob stuff for now.  Good catch though

Comment (markfields) [reply packages/runtime/container-runtime/src/blobManager.ts]: Tombstone errors thrown to (through) handle.get will include a headers object like `{ isTombstoned: true }`

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts (hunk 6 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:560]: Recovery is covered in another test (Can un-tombstone datastores by storing a handle)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 7]
```diff
@@ -524,6 +524,30 @@ export enum RuntimeHeaders {
...
+
+/**
+ * The full set of parsed header data that may be found on Runtime requests
+ */
+export interface RuntimeHeaderData {
+    wait?: boolean;
+    externalRequest?: boolean;
+    viaHandle?: boolean;
+    allowTombstone?: boolean;
+}
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/containerRuntime.ts:541]: Is this expanding our public api surface?

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:541]: Nope, see index.ts

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 8]
```diff
@@ -450,23 +462,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
+            assert.notEqual(tombstoneSuccessResponse.headers?.[TombstoneResponseHeaderKey], true, "DID NOT Expect tombstone header to be set on the response");
+
+            const summarizerResponse = await containerRuntime_resolveHandle(summarizingContainer, { url: unreferencedId });
+            assert.equal(summarizerResponse.status, 200, "Should be able to retrieve a tombstoned datastore in summarizer clients");
+            assert.notEqual(summarizerResponse.headers?.[TombstoneResponseHeaderKey], true, "DID NOT Expect tombstone header to be set on the response");
         });
 
-        // If this test starts failing due to runtime is closed errors try first adjusting `sweepTimeoutMs` above
-        itExpects("Requesting tombstoned datastores fails in summarizing container loaded before sweep timeout",
+        // SKIPPED: This test is redundant with the previous one; SweepReady calculation/timers has coverage elsewhere
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:479]: Can we just delete these tests? Also, can we just delete these tests in a separate PR? This doesn't seem like a necessary change we need to backport

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 9]
```diff
@@ -450,23 +462,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
+        itExpects.skip("Requesting tombstoned datastores fails in summarizing container loaded before sweep timeout",
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:480]: Can we just delete these tests?

Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:480]: It screws up the diff pretty bad and I wanted clear feedback on skipping them so went this way

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts [hunk 10]
```diff
@@ -23,6 +23,7 @@ interface IResponseException extends Error {
     message: string;
     code: number;
     stack?: string;
+    underlyingResponseHeaders?: { [key: string]: any; };
```
Comment (agarwal-navin) [inline packages/runtime/runtime-utils/src/dataStoreHelpers.ts:26]: Should we just call them "headers" similar to how IResponse has headers? It is after all a response exception.

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 11]
```diff
@@ -430,31 +430,39 @@ export class DataStores implements IDisposable {
...
         }
 
         if (context.tombstoned) {
+            const shouldFail = this.throwOnTombstoneLoad && !headerData.allowTombstone;
+
             // The requested data store is removed by gc. Create a 404 gc response exception.
-            const error = responseToException(createResponseError(404, "Datastore removed by gc", request), request);
+            const error = responseToException(createResponseError(
+                404,
+                "Datastore removed by gc",
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/dataStores.ts:449]: We should rename this to "Datastore was deleted". No need to mention "gc" here as it sounds like gc did something wrong :-).

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts (hunk 10 above)
Comment (markfields) [reply packages/runtime/runtime-utils/src/dataStoreHelpers.ts:26]: That's how it was, I changed it last night. Really, handle.get should expose its own shape of error and not expose the fact that internally it's making a request. But I don't want to design that right now. 

I can switch it back to headers if you care. "It's a response exception" - well there's nothing about the actual data that says that, just the TS type. Probably should have errorType but again not touching that here.

Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/dataStoreHelpers.ts:26]: I don't care that much :-). It's not a big deal as its internal.
Agree with you on handle.get() errors. We should define the error type. Hopefully with the handle work, this will get done. We can create an issue for now.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 12]
```diff
@@ -450,23 +468,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
-                    const correctErrorMessage = error.message === `Datastore removed by gc: ${unreferencedId}`;
-                    return correctErrorType && correctErrorMessage;
-                },
-                `Should not be able to retrieve a tombstoned datastore in non-summarizer clients`,
-            );
-            await assert.doesNotReject(
-                async () => requestFluidObject<ITestDataObject>(summarizingContainer, unreferencedId),
-                `Should be able to retrieve a tombstoned datastore in summarizer clients`,
-            );
+            const tombstoneErrorResponse = await containerRuntime_resolveHandle(container, { url: unreferencedId });
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:471]: Can you add comments here explaining why this fails and the next request passes.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 13]
```diff
@@ -587,21 +609,17 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
-            // since the summarizer client is the one that is created and adding the handle, the handle is not decoded
-            // and it will not result in GC detecting the reference. Also, a "gcUnknownOutboundReferences" error will
-            // be logged for the same reason.
-            const mainDataObject = await requestFluidObject<ITestDataObject>(summarizingContainer, "default");
-            mainDataObject._root.set("store", dataObject.handle);
+            const requestAllowingTombstone: IRequest = {
+                url: unreferencedId,
+                headers: { [AllowTombstoneRequestHeaderKey]: true },
+            };
+            const tombstonedObject = await requestFluidObject<ITestDataObject>(tombstoneContainer, requestAllowingTombstone);
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:620]: Why use requestFluidObject here and containerRuntime_resolveHandle above?

Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:620]: I think anywhere that was getting the handle off the data object I used RequestFluidObject rather than writing that Provider code myself. It's a little arbitrary, definitely open to input.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 14]
```diff
@@ -521,17 +536,22 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
-            assert(response.value === `Datastore removed by gc: ${unreferencedId}`);
+            let tombstoneError: Error & { code: number; underlyingResponseHeaders?: { [TombstoneResponseHeaderKey]: boolean; }; } | undefined;
+            try {
+                await handle.get();
+            } catch (error: any) {
+                tombstoneError = error
+            }
+            assert.equal(tombstoneError?.code, 404, "Tombstone error from handle.get should have 404 status code");
+            assert.equal(tombstoneError?.message, `Datastore removed by gc: ${unreferencedId}`, "Incorrect message for Tombstone error from handle.get");
+            assert.equal(tombstoneError?.underlyingResponseHeaders?.[TombstoneResponseHeaderKey], true, "Tombstone error from handle.get should include the tombstone flag");
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:550]: Is there any value in adding retrieving this object via resolveHandle after this to mimic what the actual usage would be?

**Review by agarwal-navin — APPROVED** [review]

File: packages/runtime/container-runtime/src/dataStores.ts (hunk 11 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/dataStores.ts:449]: (Pending)

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts (hunk 14 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:550]: Ok sure
